Add basic testing page for browser quicks

nhunzaker opened the pull request:
Related to #8583

---

This is just the start. I'd like to dynamically pull in versions some how, and @gaearon mentioned renaming `examples` to `fixtures`, but it accomplishes most of what we need (I think).

Next step would be to create a basic template for a fixture.

![test-page](https://cloud.githubusercontent.com/assets/590904/21265640/6af5275c-c370-11e6-9bc4-196e99b73a4a.gif)

---

@Aweary Should we start working from a branch other than master? I can't do that, but I'd be happy to adjust the base of this PR to a `browser-quirks` (or whatever) branch on the main repo.

nhunzaker commented on examples/basic/index.html:
```diff
@@ -23,8 +23,9 @@
-    <script src="../../build/react.js"></script>
-    <script src="../../build/react-dom.js"></script>
+
+    <script src="../react-loader.js"></script>
```
I've only done it for the basic example, but I could imagine doing this everywhere. We could also add an extra flag to pull in Babel.

nhunzaker commented on examples/test.html:
```diff
@@ -0,0 +1,106 @@
+            <option value="15.3.0">15.3.0</option>
+            <option value="15.2.1">15.2.1</option>
+            <option value="15.1.0">15.1.0</option>
+          </select>
```
Totally arbitrary and temporary list. What's a good way to pull in all of the version data?

Aweary replied:
It would be nice if unpkg or npm had a small API to get this information, but it doesn't look like they do. The easiest way may be to just manually keep it up to date as part of the release process. It would be nice to have something like CHANGELOG but in JSON.

vinnymac replied:
@nhunzaker If you make a request to the NPM registry and parse out the version keys, you should be able to get a similar output to the `npm view react versions` command. Here is the the response on the registry for [react](https://registry.npmjs.org/react), it contains all versions published. It may be more information than you need, but I hope it helps. (Note: You could fall back to [Yarn](https://registry.yarnpkg.com/react) if NPM was down.)

nhunzaker replied:
@vinnymac Very cool. I think tags is precisely what I want.

nhunzaker replied:
Added. We'll pull from the Github API, though this fails in IE9 because of silliness with CORS and XDomainRequest. Something to figure out later... 

Otherwise, this page looks good and works down to IE9.

Aweary commented:
> @Aweary Should we start working from a branch other than master? I can't do that, but I'd be happy to adjust the base of this PR to a browser-quirks (or whatever) branch on the main repo.

I think I can just push changes to your branch, as long as you didn't uncheck that option when opening the PR.

Aweary commented:
This is a good start 👍 How are we going to structure things at a high-level? Should tests be grouped by together in separate pages (`input.html`, `textarea.html`, `select.html`, etc.) or are we going to just keep it on a single page?

Personally, I think separate pages is the way to go since I expect this to get pretty large by the time we're done.

If we *do* use separate pages, does the `iframe` approach make sense? Would toggling between different pages work like toggling between different React versions?

nhunzaker commented:
> I think I can just push changes to your branch, as long as you didn't uncheck that option when opening the PR.

Woah neat. Good deal.

I think we should do separate pages too. We could add a dropdown just like React versions. Maybe setup a script that generates it (or do it by hand for a while).

Another part of this is documentation. Where does documentation for fixes to random browser issues live?

nhunzaker commented:
Updated the readme in `examples/inputs` to talk about the difference between `defaultValue` and `value`, and the concept of value detachment.

nhunzaker commented:
@gaearon mentioned renaming `examples` to `fixtures`. For the sake of keeping this up to date with master, I'll go ahead and make a fixtures folder and move all of this work over to that. We can always rename/remove examples at a later point.

nhunzaker commented on fixtures/README.md:
```diff
@@ -0,0 +1,14 @@
+IE Edge
+Chrome
+Firefox
+Safari
```
Curiously, I couldn't find anything about base line browser support for React. Would it be the same as Facebook? Either way, do you know what this support matrix looks like?

Otherwise, drawing from the list of ES5 compliant browsers makes the most sense to me:
http://caniuse.com/#search=ES5

nhunzaker replied:
Ah... neat. Thank you!

gaearon commented:
Excited to see work on this. Let me know if you get blocked on something!

nhunzaker commented:
@gaearon you're fast! I was just writing this up..

Added:

- Test for selects
- Test for range inputs
- Test for textareas

I also write up a section on `form.reset()`, documented the current issues with number inputs, and moved all of this work to a `fixtures` folder.

At this point, I feel good about everything. The only outstanding issues relates to fetching the version information in IE9 for the dropdown menu. 

I think a next step would be to catalog past issues. I can think of two off the top of my head:

1. Disabled buttons/inputs
2. Assignment of min/max/step before value (causing display value issues in IE)
3. Display value issues with improperly detached date inputs in iOS Safar /Android Chrome

Though I'm sure there are more. 

Still, I think these are additive. At some point, I would love to check this in so that others can contribute to it more easily.

nhunzaker commented:
And... just fixed the lint issues. Hopefully CI Passes now.

aweary commented:
@nhunzaker I added a small commit to show the values for the controlled `<textarea/>` and `<select/>`. I think we should make sure to always display the output for controlled components.

I think we need to decide on some high-level design structure before adding anything else. We want to have an accessible set of test cases that anyone can run through. It should be easy enough for a new contributor to intuitively follow. IMO the clearest way to do that is to have instructions inline.
Then we'd have a page with a set of generalized components (controlled, uncontrolled, etc.) and then below them a set of interactive test cases. That would look something like:

<img width="496" alt="screen shot 2016-12-22 at 9 38 50 am" src="https://cloud.githubusercontent.com/assets/6886061/21430850/8ba383f4-c82a-11e6-95bf-5047ead10cb7.png">


The advantages with this approach is that it makes it very, very easy for a contributor to open this page and work from a central place to test their changes. We could even have little checkboxes that people can use to track which tests they've done every time their refresh/check changes.

The downside is that our component structure increases in complexity. Each individual test case would likely need to maintain its own state and all that. If we did it this way, we'd likely want to start thinking about a better structure (instead of a bunch of inlined components)

Alternatively we could have them listed in each individual README. The advantage there is that its probably easier to maintain and easier to read over. The downside is that it makes it slightly harder to track what you're doing and requires you to have two pages open at once.

What do you think?

nhunzaker commented:
I like the first approach. Especially if we can make it available online. Each file would be a stand alone issue reproducer, making it easier to produce and avoids possibly missing context.

So let's start making cases?

aweary commented:
Should we move away from loading static HTML files and provide a server that users can run? It would make future deployment efforts easier. Then we could actually separate stuff out into different files and organize a bit better. And I think we'd sidestep CORS issue too since we can just request version history from the server. 

Then users can just do `npm run fixtures` or something

gaearon commented:
I feel like there's a lot of bikeshed potential if you move into the directions of servers and many pages. I would probably go with a single page and many components. Or you can use existing infra like Create React App.

aweary commented:
I'm not saying do many individual pages, the current strategy of toggling between the different categories of components with a dropdown is a good approach. I'm just saying we can serve this file from `localhost` instead of just from `file://`. It would just be a simple webpack setup serving essentially the same thing we have now, except we wouldn't have to inline everything in a `script` tag.

> Or you can use existing infra like Create React App.

I like that idea! That actually seems like the most logical way to do it.

nhunzaker commented:
>  It would just be a simple webpack setup serving essentially the same thing we have now

Maybe we could even.... get incremental react builds!

nhunzaker commented:
I'm fine with either `file://` vs `http://`, whatever is the easiest to author and set up for new contributors. 

As for writing the tests themselves:

What I like about @aweary's decimal dropping example is that it includes specific reproduction steps for a confusing issue. Does it make sense to have a few pages for those strange edge cases, however maintain large pages by component type like we have now?

I can see a few focused examples from my past experience:

1. The `<input type=range />` hit on IE
2. Disabled button bubbling
3. Chrome/Safari backspace issues
4. Chrome/Safari DateTime value detachment issues

So pages for a particular problem. Ones that require special handling inside of React, or we don't want to lose history on.

Does this make sense?

nhunzaker commented:
Also, I'm fine punting on specific test cases in favor of the "kitchen sink" style pages until we ship this. To me, they are more of documentation than anything else. 

@aweary What does your availability look like? Would you like to own setting up a `fixtures` `create-react-app`? Otherwise I can probably get to it either tonight, or next Tuesday.

aweary commented:
@aweary I can own getting a fixtures app setup using `create-react-app` 🤘

aweary commented:
> Does it make sense to have a few pages for those strange edge cases, however maintain large pages by component type like we have now?

I think we should have large pages for each category of components like now, and just have the general "kitchen sink" playground at the top, and then have a section below that lists each test case. That way if we provide a checkbox for each one they can just run through the list quickly and verify everything works with their changes, and it's easy to just scroll the page and see if there are any cases you didn't hit.

So it would go:

# Kitchen Sink / Playground
________

aweary commented:
So I looked into getting an app setup using `create-react-app` and it's unfortunately not as straightforward as I'd hoped. Since we need to dynamically load different versions of `React` and `ReactDOM`, we need to make sure webpack doesn't bundle those. We can do that my marking them as external and making sure `React` and `ReactDOM` are available  *before* the bundle is loaded, but that means we need to eject and I think that would defeat the purpose of using `create-react-app` in the first place (we now have our own infra to maintain anyways).

gaearon commented:
You can just read them from the global. CRA doesn't force you to import React in a specific way.

```
const ReactDOM = window.ReactDOM;
const React = window.React;
const Page = createPage(React, ReactDOM);
ReactDOM.render(<Page />, el);
```

aweary commented:
I somehow didn't even think of doing that, thanks!

aweary commented:
@nhunzaker @gaearon I've pushed my commit with the application created using `create-react-app` (https://github.com/facebook/react/pull/8589/commits/99bfa2892b1c2843d6a23b520f735925967d1a46) . It works pretty well overall. There are some notes in the commits about implementation details, and I'm not 100% on the folder structure, but I wanted to get something up for review 👍

aweary commented on fixtures/README.md:
```diff
@@ -1,16 +1 @@
-### IE8 Support
-
-React 0.14 and lower support IE8 after including several required shims.
+# React Fixtures (TODO)
```
This got deleted when I merged, I'll make sure to add it back

aweary commented on fixtures/package.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "_fixtures",
```
What should we call it? `react-fixtures`?

nhunzaker replied:
👍.

aweary commented on fixtures/public/react-loader.js:
```diff
@@ -7,8 +7,8 @@
-var REACT_PATH = '../../build/react.js';
-var DOM_PATH = '../../build/react-dom.js';
+var REACT_PATH = 'react.js';
+var DOM_PATH = 'react-dom.js';
```
We need to figure out the best way to get a copy of the local `react` and `react-dom` builds.  The relative path was not working out of the box, so I ended up just copying the files from `build` manually. Maybe we should make that part of the build process, or do it before the `start` script for this app?

If you're testing locally, make sure to copy those files into here for now!

nhunzaker replied:
A `prestart` command that copies everything over is fine by me.

aweary commented on fixtures/src/components/Header.js:
```diff
@@ -0,0 +1,68 @@
+    return { version, versions };
+  },
+  componentWillMount() {
+    fetch('http://api.github.com/repos/facebook/react/tags')
```
Are we all OK with shipping this with a `fetch` polyfill so we can use this ^?

nhunzaker replied:
Fine by me. `create-react-app` comes with it already too.

nhunzaker commented:
Wow. Time to catch up! Happy new year!

aweary commented on fixtures/public/react-loader.js:
```diff
@@ -0,0 +1,49 @@
+ */
+
+var REACT_PATH = 'react.js';
+var DOM_PATH = 'react-dom.js';
```
Question got hidden as "outdated", but its not: https://github.com/facebook/react/pull/8589#discussion_r94507090

aweary commented:
@nhunzaker :tada: everything is mostly the same as far as the actual fixtures go, I just moved them into their own files and got linting to pass.

nhunzaker commented on fixtures/src/components/Header.js:
```diff
@@ -0,0 +1,68 @@
+        <div className="header-controls">
+          <label htmlFor="example">
+            <span className="sr-only">Select an example</span>
+            <select value={window.location.pathname} onChange={this.handleFixtureChange}>
```
Should this value be in initial state?

aweary replied:
Yeah, probably. The way I do it with version is to just pass it from the `window.location` and update `window.location.search` when it changes, forcing it to reload (since we need a full reload on version change).

I did the same thing here, but I guess we don't need a full reload on fixture change?

nhunzaker replied:
Reply via email didn't match up:

> Naw, we don't need to reload, but I don't know how much we want to mess
with history API support. Worth it, you think?

aweary replied:
Yeah, I already feel unsure about setting `window.location` properties directly, does that consistently cause a refresh in all supported browsers? If we don't need to do it then 👍 on avoid it

nhunzaker replied:
It sure does.

nhunzaker replied:
Oh, herm... I'll check. Be back shortly.

nhunzaker replied:
Checks out for me down to IE9 and all major browsers

nhunzaker commented on fixtures/src/components/fixtures/range-inputs/index.js:
```diff
@@ -0,0 +1,28 @@
+      <form>
+        <fieldset>
+          <legend>Controlled</legend>
+          <input type="range" value={this.state.value} onChange={this.onChange.bind(this)} />
```
no need to bind `onChange` if we're using `createClass`

nhunzaker replied:
Probably my mistake :)

nhunzaker commented:
This looks great! in #8583 I had two other items:

- [] Catalog existing fixes and outstanding issues, start building test cases
- [] Modify React flags via query strings, or some other mechanism to quickly switch between the Fiber and and standard DOM renderer.

Would you consider the first to be reasonably complete, and do we want to cover feature flags in this PR?

nhunzaker commented:
Naw, we don't need to reload, but I don't know how much we want to mess
with history API support. Worth it, you think?

On Tue, Jan 3, 2017 at 6:56 PM Brandon Dail <notifications@github.com>
wrote:

> *@aweary* commented on this pull request.
> ------------------------------
>
> In fixtures/src/components/Header.js
> <https://github.com/facebook/react/pull/8589>:
>
> > +  handleFixtureChange(event) {
> +    window.location.pathname = event.target.value;
> +  },
> +  render() {
> +    return (
> +    <header className="header">
> +      <div className="header__inner">
> +        <span className="header__logo">
> +          <img src="https://facebook.github.io/react/img/logo.svg" alt="" width="32" height="32" />
> +          React Sandbox (v{React.version})
> +        </span>
> +
> +        <div className="header-controls">
> +          <label htmlFor="example">
> +            <span className="sr-only">Select an example</span>
> +            <select value={window.location.pathname} onChange={this.handleFixtureChange}>
>
> Yeah, probably. The way I do it with version is to just pass it from the
> window.location and update window.location.search when it changes,
> forcing it to reload (since we need a full reload on version change).
>
> I did the same thing here, but I guess we don't need a full reload on
> fixture change?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/react/pull/8589>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkEON719QhtcVIoGwMCxJO6tlSkJaB2ks5rOuArgaJpZM4LPOoQ>
> .
>

nhunzaker commented:
Hey, fun! Looks like we're hitting the new `document.write` thing in Chrome:

<img width="857" alt="screen shot 2017-01-03 at 7 10 22 pm" src="https://cloud.githubusercontent.com/assets/590904/21627793/59e02cf6-d1e8-11e6-902c-b73944ec2ddc.png">

Joy! I'll fix this.

nhunzaker commented:
Well.. we don't need babel for sure anymore. I think the other thing was related to a local build. I'll add that pre-start script.

nhunzaker commented:
@aweary That detail work, yo:

<img width="446" alt="screen shot 2017-01-03 at 7 17 53 pm" src="https://cloud.githubusercontent.com/assets/590904/21627927/7845284e-d1e9-11e6-8dad-de4154235a90.png">

🔥

nhunzaker commented:
Pushed some stuff:

- remove babel include from react-loader
- Copy local build before starting
- Add test prompt message when no fixture is selected
- Remove extraneous binds from some fixtures

gaearon commented:
What is the status here?

nhunzaker commented:
@gaearon doing some browser testing. I'll report back shortly.

aweary commented:
> What is the status here?

We've setup an application using `create-react-app` (see https://github.com/facebook/react/pull/8589/commits/99bfa2892b1c2843d6a23b520f735925967d1a46 and the handful of commits @nhunzaker has made since) with the ~same content we had before. Just wanting to see if you had any feedback on the overall app structure + implementation. No rush 👍 

> Would you consider the first to be reasonably complete, and do we want to cover feature flags in this PR?

@nhunzaker I think we can implement feature flags in this PR and continue documenting test cases in #8583. It might be good to merge this soon and then add the other test cases in future PRs, so this doesn't get too big.

gaearon commented:
Can we move descriptions of test cases inline inside the page? So that you can just fire it up and walk through the whole thing without switching applications.

aweary commented:
@gaearon currently this PR just contains the general kitchen sink components. We still need to create the individual test cases.

I'm thinking we can have a component like `<FixtureTestCase />` that lets you define a set of steps and provides an interface for marking the case as "tested" so you can track progress as you go through the page.

```jsx
<FixtureTestCase title="Foo" steps={steps}>
  // components for test case go here
</FixtureTestCase>

gaearon commented:
>We still need to create the individual test cases, if that's what you mean?

Yea.

nhunzaker commented:
Did some cross browser checking/fixing. Also simplified styles into a single file, we needed the input form styles in the textarea example too. 

So the outstanding tasks:

- [ ] Individual test cases
- [ ] Move descriptions of test cases inline inside the page? 

@gaearon would you say that this is correct?

nhunzaker commented:
Fixed lint issues. Though I think I'm hitting this locally:

```
Cannot find module 'eslint-config-postcss/es5'
Referenced from: <...>/react/fixtures/node_modules/autoprefixer/package.json
Error: Cannot find module 'eslint-config-postcss/es5'
```

So let's see what CI says. Curious why this isn't throwing on CI though.

gaearon commented:
I just merged a folder named `fixtures` in an unrelated commit (for manual testing of build systems). Could you please move its content into `fixtures/build` and make this `fixtures/dom` or something?

nhunzaker commented:
Done. I've also updated the readme with setup instructions.

nhunzaker commented:
As for inlining the test case descriptions. Every description other than text inputs was pretty low value. Just a quick line that said something like "This fixture tests <textarea />".

However the input README is pretty involved, and talks through some important concepts.

@gaearon did you envision that we would put that information within the text case? Or rather that we'd place instructions for how to use the text fixture along side the text case? Where should we put something like [`fixtures/inputs/README.md`](https://github.com/nhunzaker/react/blob/3d6a63d3c62eefab3f1534e43af7c8451def5e1f/fixtures/dom/src/components/fixtures/text-inputs/README.md)

nhunzaker commented on .gitignore:
```diff
@@ -19,6 +19,8 @@ docs/downloads/*.zip
+fixtures/dom/public/react.js
 test/the-files-to-test.generated.js
 *.log*
 chrome-user-data
```
Thanks, 👍

aweary commented:
@nhunzaker I think we could have the READMEs explain high-level concepts and implementation details that are good to know, but not crucial for testing. Then we inline instructions that are just to-the-point directions on how to validate expected behavior. That way we don't include too much detail on the fixtures page, but its still available somewhere.

nhunzaker commented:
@aweary that would be my preference as well.

gaearon commented:
Sounds good. Really, the only thing necessary is that it's easy both for the team and the casual contributors to understand what they should test and how it should behave.

jquense commented:
:wave: just finding this PR. I'd be happy to help out if there is still work to be done :)

jquense commented on fixtures/dom/src/components/fixtures/text-inputs/README.md:
```diff
@@ -0,0 +1,65 @@
+There are a couple of important concepts to be aware of when working on text
+inputs in React.
+
+## `defaultValue` vs `value`
```
perhaps worth noting that React explicitly supports changing the value of an input imperatively via JavaScript. This isn't an obvious invariant when making changing to input code, but required to probably support autofilling, password managers, and general DOM interop (unfortunately).

gaearon replied:
Good catch. We’ll want to add a fixture for this.

jquense commented on fixtures/dom/src/components/fixtures/text-inputs/index.js:
```diff
@@ -0,0 +1,66 @@
+    // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input
+    let types = [
+      'text', 'email', 'number', 'url', 'tel',
+      'color', 'date', 'datetime-local',
```
AFAIK  `datetime-local` is still supported as `datetime` in some browsers, I'm not sure if that's worth adding or noting tho.

aweary commented:
> 👋 just finding this PR. I'd be happy to help out if there is still work to be done :)

👋 @jquense we still need to write fixtures for all the specific test cases. Right now we just have the "kitchen sink" pages with controlled/uncontrolled variants of a few input types.

I mentioned in https://github.com/facebook/react/pull/8589#issuecomment-270267194 that we might want to write a component that makes it easy to create new fixtures that can be marked as completed (for users validating behavior when making changes)

Any input (heh) you have on scenarios we should be explicitly testing would be amazing! I'll start getting a list together soon and add it to the top comment in this PR.

gaearon commented:
>Done. I've also updated the readme with setup instructions.
 
Actually I think `fixtures/build` disappeared. This name is probably in gitignore.
You can name it `fixtures/packaging` instead.

gaearon commented:
After fixing https://github.com/facebook/react/pull/8589#issuecomment-271385313 feel free to merge this when you feel like it. The PR is getting quite large.

gaearon commented:
In addition to fixing https://github.com/facebook/react/pull/8589#issuecomment-271385313 you'll also need to fix the build fixtures to resolve paths correctly since their relative paths to the build folder are now wrong. You can try to go through the README to see the problem.

nhunzaker commented:
@gaearon: Sorry for going quiet. Busy day! I should be able to get to this feedback tonight.

aweary commented:
@nhunzaker I'm working on it right now, I'll get it fixed and then merged 😃

gaearon commented:
Thanks @aweary!

nhunzaker commented:
@aweary True friend ❤️

aweary commented on fixtures/packaging/webpack/config.js:
```diff
@@ -1,9 +1,11 @@
   },
   resolve: {
-    root: '../../build/packages',
+    root: path.resolve('../../../build/packages/'),
```
@gaearon looks like [webpack requires](https://webpack.github.io/docs/configuration.html#resolve-root) `resolve.root` to be absolute, had to make this change for the webpack packaging fixtures to build

gaearon replied:
Got it, thanks

aweary commented:
Made the required changes, just waiting on CI to approve and will merge after 👍
